README.md
# Pricing cards

[![pub package](https://img.shields.io/pub/v/pricing_cards?logo=dart&label=stable&style=flat-square)](https://pub.dev/packages/pricing_cards) [![GitHub license](https://img.shields.io/github/license/f2acode/pricing_cards?style=flat-square)](https://github.com/f2acode/pricing_cards/blob/master/LICENSE)

Awesome pricing cards with custom style options for your flutter project.

You can give as **much cards you want** to this and **personalize the style of each one**.

## Example

The example with all the parameters:

<img src='https://raw.githubusercontent.com/f2acode/pricing_cards/master/example/flutter_01.png' width=300>

The code of this:

```dart
Column(
    mainAxisAlignment: MainAxisAlignment.center,
    children: [
        PricingCards(
            pricingCards: [
                PricingCard(
                title: 'Monthly',
                price: '\$ 9.99',
                subPriceText: '\/mo',
                billedText: 'Billed monthly',
                onPress: () {
                    // make your business
                },
                ),
                PricingCard(
                title: 'Monthly',
                price: '\$ 59.99',
                subPriceText: '\/mo',
                billedText: 'Billed anually',
                mainPricing: true,
                mainPricingHighlightText: 'Save money',
                onPress: () {
                    // make your business
                },
                )
            ],
        ),
        SizedBox(height: 30),
        PricingCards(
            pricingCards: [
                PricingCard(
                title: 'Monthly',
                price: '\$ 9.99',
                subPriceText: '\/mo',
                billedText: 'Billed monthly',
                onPress: () {
                    // make your business
                },
                cardColor: Colors.green,
                priceStyle: TextStyle(
                    fontSize: 32,
                    fontWeight: FontWeight.bold,
                    color: Colors.white,
                ),
                titleStyle: TextStyle(
                    fontSize: 16,
                    color: Colors.white,
                ),
                billedTextStyle: TextStyle(
                    fontSize: 12,
                    color: Colors.white,
                ),
                subPriceStyle: TextStyle(
                    fontSize: 12,
                    color: Colors.white,
                ),
                cardBorder: RoundedRectangleBorder(
                    side: BorderSide(color: Colors.red, width: 4.0),
                    borderRadius: BorderRadius.circular(8.0),
                ),
                ),
                PricingCard(
                title: 'Monthly',
                price: '\$ 59.99',
                subPriceText: '\/mo',
                billedText: 'Billed anually',
                mainPricing: true,
                mainPricingHighlightText: 'Save money',
                onPress: () {
                    // make your business
                },
                cardColor: Colors.blue,
                priceStyle: TextStyle(
                    fontSize: 32,
                    fontWeight: FontWeight.bold,
                    color: Colors.white,
                ),
                titleStyle: TextStyle(
                    fontSize: 16,
                    color: Colors.white,
                ),
                billedTextStyle: TextStyle(
                    fontSize: 12,
                    color: Colors.white,
                ),
                subPriceStyle: TextStyle(
                    fontSize: 12,
                    color: Colors.white,
                ),
                cardBorder: RoundedRectangleBorder(
                    side: BorderSide(color: Colors.red, width: 4.0),
                    borderRadius: BorderRadius.circular(8.0),
                ),
                )
            ],
        ),
    ],
)
```

## Help this project

- Give an 🌟 on Github;
- Contributions are very welcome! ❤️


CHANGELOG.md
## 0.0.8

- Fix duplicated Badge import

## 0.0.7

- Improve `dartdoc`;
- Provide a better (and longer) description of this package;
- Improve README.

## 0.0.6

- Convert into a package

## 0.0.5

- Fix repo

## 0.0.4

- Fix image on README

## 0.0.3

- Improve README

## 0.0.2

- Improve README

## 0.0.1

- Initial release


test/pricing_cards_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('pricing_cards');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:pricing_cards_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/pricing_cards.dart
/// Awesome flexible pricing cards with custom style options for you Flutter project
library pricing_cards;

import 'package:badges/badges.dart';
import 'package:flutter/material.dart' hide Badge;

/// The model to define data input for [PricingCards]
class PricingCard {
  /// Creates a new instance of a card
  const PricingCard({
    required this.title,
    required this.price,
    required this.subPriceText,
    required this.billedText,
    this.onPress,
    this.mainPricing = false,
    this.mainPricingHighlightText = '',
    this.priceStyle = const TextStyle(
      fontSize: 24,
      fontWeight: FontWeight.w600,
    ),
    this.billedTextStyle = const TextStyle(
      fontSize: 12,
      color: Colors.grey,
    ),
    this.cardColor = Colors.white,
    this.titleStyle,
    this.cardBorder,
    this.subPriceStyle,
  });

  /// The top text
  final String title;

  /// The main big text
  final String price;

  /// The small text under the price
  final String subPriceText;

  /// The (possibly) smooth text on the bottom
  final String billedText;

  /// The function called when the user press this card
  final void Function()? onPress;

  /// Defines this card as main so it will have a special [cardBorder] and a badge
  final bool mainPricing;

  /// The text that will appear above the card if [mainPricing] is true
  final String mainPricingHighlightText;

  /// The [price] style
  final TextStyle? priceStyle;

  /// The [billedText] style
  final TextStyle? billedTextStyle;

  /// The [subPriceText] style
  final TextStyle? subPriceStyle;

  /// The [title] style
  final TextStyle? titleStyle;

  /// The border style of this
  final RoundedRectangleBorder? cardBorder;

  /// The background color of this
  final Color? cardColor;
}

/// A widget list with one or more [PricingCards]
class PricingCards extends StatelessWidget {
  /// Creates a widget list of [PricingCard]
  ///
  /// This list wraps accordingly to the width of the parent.
  const PricingCards({
    Key? key,
    required this.pricingCards,
  }) : super(key: key);

  /// The list of this, you can give this as much [PricingCard] you want
  final List<PricingCard> pricingCards;

  @override
  Widget build(BuildContext context) {
    return Wrap(
      alignment: WrapAlignment.center,
      children: pricingCards.map((pc) {
        final pricingCard = Row(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Card(
              color: pc.cardColor,
              shape: pc.mainPricing
                  ? pc.cardBorder ??
                      RoundedRectangleBorder(
                        side: BorderSide(color: Colors.red, width: 4.0),
                        borderRadius: BorderRadius.circular(8.0),
                      )
                  : null,
              child: Container(
                padding: EdgeInsets.all(16),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Text(
                      pc.title,
                      style: pc.titleStyle,
                    ),
                    SizedBox(height: 16),
                    Text(
                      pc.price,
                      style: pc.priceStyle,
                    ),
                    Text(
                      pc.subPriceText,
                      style: pc.subPriceStyle,
                    ),
                    SizedBox(height: 16),
                    Text(
                      pc.billedText,
                      style: pc.billedTextStyle,
                    ),
                  ],
                ),
              ),
            ),
          ],
        );

        return InkWell(
          onTap: pc.onPress,
          child: pc.mainPricing
              ? Badge(
                  alignment: Alignment.center,
                  badgeContent: Text(
                    pc.mainPricingHighlightText,
                    style: TextStyle(color: Colors.white),
                  ),
                  shape: BadgeShape.square,
                  position: BadgePosition.topEnd(top: -10),
                  borderRadius: BorderRadius.circular(16),
                  toAnimate: false,
                  child: pricingCard,
                  badgeColor: Colors.red,
                  padding: EdgeInsets.symmetric(horizontal: 12, vertical: 4),
                )
              : pricingCard,
        );
      }).toList(),
    );
  }
}


android/src/main/kotlin/com/example/pricing_cards/PricingCardsPlugin.kt
package com.example.pricing_cards

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** PricingCardsPlugin */
class PricingCardsPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pricing_cards")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:pricing_cards/pricing_cards.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              PricingCards(
                pricingCards: [
                  PricingCard(
                    title: 'Monthly',
                    price: '\$ 9.99',
                    subPriceText: '\/mo',
                    billedText: 'Billed monthly',
                    onPress: () {
                      // make your business
                    },
                  ),
                  PricingCard(
                    title: 'Monthly',
                    price: '\$ 59.99',
                    subPriceText: '\/mo',
                    billedText: 'Billed anually',
                    mainPricing: true,
                    mainPricingHighlightText: 'Save money',
                    onPress: () {
                      // make your business
                    },
                  )
                ],
              ),
              SizedBox(height: 30),
              PricingCards(
                pricingCards: [
                  PricingCard(
                    title: 'Monthly',
                    price: '\$ 9.99',
                    subPriceText: '\/mo',
                    billedText: 'Billed monthly',
                    onPress: () {
                      // make your business
                    },
                    cardColor: Colors.green,
                    priceStyle: TextStyle(
                      fontSize: 32,
                      fontWeight: FontWeight.bold,
                      color: Colors.white,
                    ),
                    titleStyle: TextStyle(
                      fontSize: 16,
                      color: Colors.white,
                    ),
                    billedTextStyle: TextStyle(
                      fontSize: 12,
                      color: Colors.white,
                    ),
                    subPriceStyle: TextStyle(
                      fontSize: 12,
                      color: Colors.white,
                    ),
                    cardBorder: RoundedRectangleBorder(
                      side: BorderSide(color: Colors.red, width: 4.0),
                      borderRadius: BorderRadius.circular(8.0),
                    ),
                  ),
                  PricingCard(
                    title: 'Monthly',
                    price: '\$ 59.99',
                    subPriceText: '\/mo',
                    billedText: 'Billed anually',
                    mainPricing: true,
                    mainPricingHighlightText: 'Save money',
                    onPress: () {
                      // make your business
                    },
                    cardColor: Colors.blue,
                    priceStyle: TextStyle(
                      fontSize: 32,
                      fontWeight: FontWeight.bold,
                      color: Colors.white,
                    ),
                    titleStyle: TextStyle(
                      fontSize: 16,
                      color: Colors.white,
                    ),
                    billedTextStyle: TextStyle(
                      fontSize: 12,
                      color: Colors.white,
                    ),
                    subPriceStyle: TextStyle(
                      fontSize: 12,
                      color: Colors.white,
                    ),
                    cardBorder: RoundedRectangleBorder(
                      side: BorderSide(color: Colors.red, width: 4.0),
                      borderRadius: BorderRadius.circular(8.0),
                    ),
                  )
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
